dll/win32/advapi32/reg/hkcr.c

```diff
@@ -1014,3 +1014,68 @@ EnumHKCRValue(
 
     return ErrorCode;
 }
+
+/* HKCR version of RegQueryInfoKeyW */
+LONG
+WINAPI
+QueryInfoHKCRKey(
+    _In_ HKEY hKey,
+    _Out_writes_to_opt_(*lpcchClass, *lpcchClass + 1) LPWSTR lpClass,
+    _Inout_opt_ LPDWORD lpcchClass,
+    _Reserved_ LPDWORD lpReserved,
+    _Out_opt_ LPDWORD lpcSubKeys,
+    _Out_opt_ LPDWORD lpcbMaxSubKeyLen,
+    _Out_opt_ LPDWORD lpcbMaxClassLen,
+    _Out_opt_ LPDWORD lpcValues,
+    _Out_opt_ LPDWORD lpcbMaxValueNameLen,
+    _Out_opt_ LPDWORD lpcbMaxValueLen,
+    _Out_opt_ LPDWORD lpcbSecurityDescriptor,
+    _Out_opt_ PFILETIME lpftLastWriteTime)
+{
+    HKEY PreferredKey, FallbackKey;
+    LONG retval, err;
+    DWORD OtherSubKeys = 0, OtherMaxSub = 0, OtherMaxClass = 0;
+    DWORD OtherValues = 0, OtherMaxName = 0, OtherMaxVal = 0;
+
+    ASSERT(IsHKCRKey(hKey));
+
+    /* Remove the HKCR flag while we're working */
+    hKey = (HKEY)(((ULONG_PTR)hKey) & ~0x2);
+
+    retval = GetFallbackHKCRKey(hKey, &FallbackKey, FALSE);
+    if (retval == ERROR_SUCCESS)
+    {
+        retval = RegQueryInfoKeyW(FallbackKey, lpClass, lpcchClass, lpReserved,
+                                  &OtherSubKeys, &OtherMaxSub, &OtherMaxClass,
+                                  &OtherValues, &OtherMaxName, &OtherMaxVal,
+                                  lpcbSecurityDescriptor, lpftLastWriteTime);
+        if (FallbackKey != hKey)
+            RegCloseKey(FallbackKey);
+    }
+
+    err = GetPreferredHKCRKey(hKey, &PreferredKey);
+    if (err == ERROR_SUCCESS)
+    {
+        err = RegQueryInfoKeyW(PreferredKey, lpClass, lpcchClass, lpReserved,
+                               lpcSubKeys, lpcbMaxSubKeyLen, lpcbMaxClassLen,
+                               lpcValues, lpcbMaxValueNameLen, lpcbMaxValueLen,
+                               lpcbSecurityDescriptor, lpftLastWriteTime);
+        if (PreferredKey != hKey)
+            RegCloseKey(PreferredKey);
+    }
+
+    if (lpcSubKeys)
+        *lpcSubKeys = (err ? 0 : *lpcSubKeys) + OtherSubKeys;
```

Comment:
```suggestion
        *lpcSubKeys = (err ? ERROR_SUCCESS : *lpcSubKeys) + OtherSubKeys;
```
+ same in the other lines (not going to repeat).

Question: So this also means that these `lpcSubKeys` and co. pointers are also used as input? If so, what about using `_Inout_opt_` for annotation?

Reply:
They are not input.

OtherSubKeys is the count from HKLM.

lpcSubKeys is temporarily used to store the count from HKCU.

The output is the sum of the two counts.